### PR TITLE
[hail/ptypes] copy based on ptype and requiredeness

### PR DESCRIPTION
**Pull request by akotlar**
Remaining:
### Non-staged code
PR here: https://github.com/hail-is/hail/pull/7880 . Effectively same interface, but with a fix for the override-def-with-optional-arguments issue.

### Remaining question:

edit: From our conversion this afternoon (1/14 @4pm), I think these questions are potentially answered: be consistent, may change.

#### Fundamental Types
I've so far settled on copyFromType requiring that the sourceType is of the same type as the type the method is being called on (for instance if this isOfType PBaseStruct, then sourceType isOfType PBaseStruct), but only checks the elements, fields' fundamentalTypes. This may be too loose: I do not fully understand the architecture of fundamentalTypes and so want to clarify this point.

#### Top-level non-missing semantics
For non-collection types (those that don't have nested fields), I expect that the function has an address pointing to data. So, for instance, in `val ptyp =PArray(PInt32())`, I expect `ptyp.copyFromType(...)` to get an address pointing to data. The inspiration from this came from RVB, which requires the same (first item in stack can't be null, but the nth-1 item can be null). So it is the responsibility of collection types to check `is*Missing`.

### CODE REVIEW

**Comment by akotlar** [comment]
after discussion with @tpoterba pausing on this until the PContainer interface is cleaned up. Aiming to finish this up next week.

**Comment by akotlar** [comment]
This is at a place where people could look.

Currently plan is to implement storeShallow on every PType separately (could also keep all in PType for convenience).

Majority of functionality: 

# TODO:

- [ ] Deep copying isn't really implemented
- [ ] Implement for things that are not PCanonicalArray or PArrayBackedContainer
- [ ] More tests

cc @patrick-schultz, @catoverdrive, @danking for feedback

File: hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala [hunk 1]
```diff
@@ -110,6 +120,9 @@ trait PArrayBackedContainer extends PContainer {
   def loadElement(aoff: Long, length: Int, i: Int): Long =
     arrayRep.loadElement(aoff, length, i)
 
+  def loadElementAddress(aoff: Code[Long], length: Code[Int], i: Code[Int]): Code[Long] =
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala]: needed to be called something other than loadElement, because scala cannot distinguish between Code[Region] and Code[Long] types in the argument list, and besides this (arity, other argument types), this method is identical to loadElement(region: Code[Region] ...), leading to type check errors.

Later on, in the PR corresponding to #7826, the region-bearing methods will go away.
references https://github.com/hail-is/hail/issues/7829

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 2]
```diff
@@ -248,5 +272,75 @@ abstract class PBaseStruct extends PType {
     }
   }
 
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcStructAddress: Code[Long],
+    allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PBaseStruct])
+
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+
+    if(this.fields == sourceType.fields) {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: I am not checking fundamentalType here, but am below. Primarily because I am still not entirely certain whether elements/fields should be copyable by their fundamentalType identity rather than their own identity.

In the same vein, if we do want to allow fundamentalType identity, would we want a typesFundamental instance attribute on this class?

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 3]
```diff
@@ -367,4 +383,79 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
+    Region.storeAddress(destOffset, valueAddress)
+  }
+
+  // semantically this function expects a non-null sourceAddress, and by that property, this function results in a non-null value
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArray])
+
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val sourceElementType = sourceType.elementType.fundamentalType
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: same question as above regarding checking equality on fundamentalTypes, note below my shallow copy check does use the fundamental version of src/dest element ptypes, not the element ptypes themselves.

It's also a cheaper check here, and 2 lines to change it back.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala [hunk 4]
```diff
@@ -50,10 +52,11 @@ final case class PCanonicalStruct(fields: IndexedSeq[PField], required: Boolean
   }
 
   val missingIdx = new Array[Int](size)
+  // sets missingIdx indices
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala]: I found this confusing. In a later PR I would like to refactor this. This pattern of mutating a passed array is unnecessary indirection, and a bit unusual for our codebase (in which functions are idempotent).

**Comment by akotlar** [comment]
@tpoterba I'm running low on time today, need to move over to stats. This is pretty close to being ready to go however. Remaining items: need to implement unstaged versions of 

Please note that for methods that are not implicitly or explicitly tested, I am not entirely certain of correctness (for instance PBinary). So if you see something stupid, know that I haven't entirely vetted it.

Remaining questions:

### Fundamental Types
I've so far settled on copyFromType requiring that the sourceType is of the same type as the type the method is being called on (for instance if this isOfType PBaseStruct, then sourceType isOfType PBaseStruct), but only checks the elements, fields' fundamentalTypes. This may be too loose: I do not fully understand the architecture of fundamentalTypes and so want to clarify this point.

In order to make switching back/forth from the fundamentalType versions, I set variables srcElementType/srcFieldType and dstElementType/dstFieldType and use those variables instead of elementType, field.type directly

File: hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala [hunk 5]
```diff
@@ -164,7 +167,21 @@ trait PArrayBackedContainer extends PContainer {
...
-  override def unsafeOrdering: UnsafeOrdering = unsafeOrdering(this)
+  override def unsafeOrdering: UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering =
+    arrayRep.unsafeOrdering(rightType)
+
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArrayBackedContainer])
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala]: We could be more general here, allow copying with a PArray as well.

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 6]
```diff
@@ -248,5 +272,76 @@ abstract class PBaseStruct extends PType {
...
+
+    assert(sourceType.size == this.size)
+
+    val sourceFundamentalTypes = sourceType.fields.map(_.typ.fundamentalType)
+    if(this.fields.map(_.typ.fundamentalType) == sourceFundamentalTypes) {
+      if(!forceDeep) {
+        return srcStructAddress
+      }
+
+      if(sourceFundamentalTypes.forall(_.isPrimitive)) {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: this is a bit wasteful, could implement in one loop, but I think this fits our typical style better

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 7]
```diff
@@ -367,4 +375,82 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
   }
+
+  override def storeShallowAtOffset(dstAddress: Code[Long], valueAddress: Code[Long]): Code[Unit] =
+    Region.storeAddress(dstAddress, valueAddress)
+
+  override def storeShallowAtOffset(dstAddress: Long, valueAddress: Long) {
+    Region.storeAddress(dstAddress, valueAddress)
+  }
+
+  // semantically this function expects a non-null sourceAddress, and by that property, this function results in a non-null value
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: This doesn't seem like a well-defined clarification in our system, since addresses cannot be null. It's like saying:

```scala
// this function expects a non-null value for 'i'
def doubleIt(i: Int): Int = i * 2
```

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 8]
```diff
@@ -367,4 +375,82 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long],
+  allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArray])
+
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val sourceElementType = sourceType.elementType.fundamentalType
+    val destElementType = this.elementType.fundamentalType
+
+    if (sourceElementType != destElementType) {
+      assert(sourceElementType isOfType destElementType)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: I find it somewhat hard to follow the control flow here, in part due to checking negatives. I think the following would be clearer:

```scala
    if (sourceElementType == destElementType) {
      if (forceDeep)
        if(sourceElementType.isPrimitive)
          return this.copyFrom(mb, region, srcAddress)
        else
          return <copy and fixup>
      return srcAddress
    }
```

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala [hunk 9]
```diff
@@ -13,7 +13,35 @@ class PCanonicalBinary(val required: Boolean) extends PBinary {
...
+      return sourceAddress
+    }
+
+    assert(this isOfType sourcePType)
+
+    val dstAddress = mb.newField[Long]
+    val length = mb.newLocal[Int]
+
+    // srcAddress must point to data by copyFromType semantics, so no runtime null-check needed
+    if(this.required > sourcePType.required) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala]: this isn't a downcast, so no need to check here. as you point out, `copyFromType` can only be called with present values.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala (hunk 4 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala]: fixed in the ptypes parser PR, as you know.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala (hunk 7 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: I was thinking about null pointers at this time, is this point worth clarifying?

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 10]
```diff
@@ -248,5 +272,76 @@ abstract class PBaseStruct extends PType {
...
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcStructAddress: Code[Long],
+    allowDowncast: Boolean = false, forceDeep: Boolean = false): Code[Long] = {
+    assert(srcPType.isInstanceOf[PBaseStruct])
+
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+
+    assert(sourceType.size == this.size)
+
+    val sourceFundamentalTypes = sourceType.fields.map(_.typ.fundamentalType)
+    if(this.fields.map(_.typ.fundamentalType) == sourceFundamentalTypes) {
```
[the remark this answers is not in the document]
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: Would PStruct(fields: Seq("1" -> 2)) == PNonCanonicalStruct(fields: Seq("1"->2))?

File: hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala [hunk 11]
```diff
@@ -164,7 +167,28 @@ trait PArrayBackedContainer extends PContainer {
...
-  override def unsafeOrdering: UnsafeOrdering = unsafeOrdering(this)
+  override def unsafeOrdering: UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering =
+    arrayRep.unsafeOrdering(rightType)
+
+  override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
+    assert(srcPType.isInstanceOf[PArrayBackedContainer])
+    println(s"The type is ${this.arrayRep.fundamentalType}")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala]: stray print

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 12]
```diff
@@ -367,4 +375,109 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
+    val dstAddress = mb.newField[Long]
+    Code(
+      dstAddress := this.copyFrom(mb, region, srcArrayAddress),
+      this.fastDeepCopyLoop(mb, region, dstAddress),
+      dstAddress
+    )
+  }
+
+  def fastDeepCopyLoop(mb: MethodBuilder, region: Code[Region], dstAddress: Code[Long]): Code[Unit] = {
+    if(this.elementType.fundamentalType.isPrimitive) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: should be `containsPointers`

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 13]
```diff
@@ -248,5 +272,105 @@ abstract class PBaseStruct extends PType {
...
+    val dstAddress = mb.newField[Long]
+    Code(
+      dstAddress := this.copyFrom(mb, region, srcStructAddress),
+      this.fastDeepCopyLoop(mb, region, dstAddress),
+      dstAddress
+    )
+  }
+
+  def fastDeepCopyLoop(mb: MethodBuilder, region: Code[Region], dstStructAddress: Code[Long]): Code[Unit] = {
+    var loop: Code[Unit] = Code._empty
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: wouldn't call this loop, since it's not generating a loop.

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 14]
```diff
@@ -248,5 +272,105 @@ abstract class PBaseStruct extends PType {
...
+
+    var i = 0
+    while(i < this.size) {
+      val dstFieldType = this.fields(i).typ.fundamentalType
+      if(!dstFieldType.isPrimitive) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: containsPointers

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala [hunk 15]
```diff
@@ -1,10 +1,11 @@
 package is.hail.expr.types.physical
 
+import is.hail.annotations.Region
 import is.hail.asm4s._
 import is.hail.expr.ir.EmitFunctionBuilder
 import is.hail.variant.Genotype
 
-final case class PCanonicalCall(required: Boolean) extends PCall {
+final case class PCanonicalCall(required: Boolean = false) extends PCall {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala]: why this change?

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala [hunk 16]
```diff
@@ -56,4 +57,18 @@ final case class PCanonicalCall(required: Boolean) extends PCall {
...
     }
 
+    override def storeShallowAtOffset(dstAddress: Code[Long], valueAddress: Code[Long]): Code[Unit] =
+      this.representation.storeShallowAtOffset(dstAddress, valueAddress)
+
+    override def storeShallowAtOffset(dstAddress: Long, valueAddress: Long) {
+      this.representation.storeShallowAtOffset(dstAddress, valueAddress)
+    }
+
+    override def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala]: I think that we can `copyFromType/StoreShallowXX` should be implemented by ComplexType to call the `representation`'s methods. This is basically dead code as existing here, since PCAnonicalCall is not a fundamental type and we only call the copy stuff on funadmental types.

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala [hunk 17]
```diff
@@ -304,4 +305,47 @@ abstract class PType extends BaseType with Serializable with Requiredness {
...
+    this.fundamentalType match {
+      case _: PBoolean | _: PInt32 | _: PInt64 | _: PFloat32 | _: PFloat64 =>
+        srcAddress
+      case ft => throw new UnsupportedOperationException("Unknown fundamental type: " + ft)
+    }
+  }
+  def copyFromType(region: Region, srcPType: PType, srcAddress: Long): Long =
+    this.copyFromType(region, srcPType, srcAddress, false)
+
+  def storeShallowAtOffset(dstAddress: Code[Long], srcAddress: Code[Long]): Code[Unit] = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: why is this implemented on PType and not on these types?

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala (hunk 15 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalCall.scala]: Because it looked like a mistake. All of our other concrete PTypes are either required = false by default on their only constructor, or have a required = false constructor.

This is the only case class that doesn't follow that pattern.

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala (hunk 13 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: I'm a bit confused by what you mean. It is, the body of the function is just a loop over elements/fields, potentially with recursion (hence deep)

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala (hunk 17 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: Just to keep the implementation less verbose, seemed to be something that PType was used for (similar to `setRequired`, `isNumeric`, `canonical`)

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala (hunk 13 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: I think you mean that the emitted code is unrolled, that’s true but loop unrolling feels like an implementation detail (for instance the interpreted code will definitely emit a loop, because the compiler doesn’t statically know size, although  the JIT may figure it out at runtime and unroll it)

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: Happy to change the name however

**Comment by akotlar** [comment]
Tim should be ready for another look, merged master
